### PR TITLE
Add a tutorial for text sentiment analysis.

### PR DESCRIPTION
**Pull request by zhangguanheng66**
Add a tutorial for text sentiment analysis. The analysis uses the TextClassification datasets added in https://github.com/pytorch/text/pull/550

### CODE REVIEW

**Comment by cpuhrsch** [comment]
Please try using EmbeddingBag, it should be much faster :)

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+   "metadata": {},
+   "source": [
+    "## Functions used to generate batch"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Functions generate_batch() is used to generate data batches and offsets, which are compatible with EmbeddingBag. EmbeddingBad is efficient to process a sequence of tensors with different lengths so that no padding is required here. This is faster than the origin method depending on TorchText Iterator and Batch."
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: I think you mean "EmbeddingBag" :)

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_batch(batch):\n",
+    "\n",
+    "    def generate_offsets(data_batch):\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: I think this needs a bit more explanation. It's not necessarily that trivial and also links together with EmbeddingBag. Since EmbeddingBag is so much more efficient than Embedding I think it's worth expanding on both.

File: docs/tutorials/text_sentiment_ngrams.ipynb (hunk 1 above)
Comment (cpuhrsch) [reply docs/tutorials/text_sentiment_ngrams.ipynb]: Can you expand a bit on the "original method depending on TorchText Iterator" and why it's faster?

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import random\n",
+    "def split_dataset(dataset, split_ratio=0.7):\n",
+    "    data = [dataset[i] for i in range(len(dataset))]\n",
+    "    random.shuffle(data)\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: We have [random_spit](https://pytorch.org/docs/master/_modules/torch/utils/data/dataset.html#random_split) in data.utils for this - This still works, but I think it's more consistent since we're using torch.utils.data.Dataset after all.

**Comment by cpuhrsch** [comment]
Some of this code is Python3 only as far as I can tell. It might be worth running this with Python2.

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "from torchtext.data.utils import ngrams_iterator\n",
+    "\n",
+    "ag_news_label = {1 : \"World\",\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: I'm wondering if we want to add "LABELS" field to torchtext.data.text_classification so that users can do torchtext.data.LABELS['AG_NEWS'][label]

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+    "                 2 : \"Sports\",\n",
+    "                 3 : \"Business\",\n",
+    "                 4 : \"Sci/Tec\"}\n",
+    "\n",
+    "vocab = train_dataset._vocab\n",
+    "\n",
+    "def predict_label(text_string, ngrams):\n",
+    "    model.eval()\n",
+    "    tokenizer = get_tokenizer(\"basic_english\")\n",
+    "    ngrams_tokens = ngrams_iterator(tokenizer(text_string), ngrams)\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: you call ngrams_iterator twice

you also don't need to create offsets if you just pass in a single line.

Look at the predict.py example to see how it's done there.

**Comment by cpuhrsch** [comment]
Try to make sure the code snippets look congruent with the example code you reference at the end.

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+    "Use the best model so far and test a golf news."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: This is unused as far as I can tell

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+    "from torchtext.data.utils import ngrams_iterator\n",
+    "from torchtext.data.utils import get_tokenizer\n",
+    "\n",
+    "ag_news_label = {1 : \"World\",\n",
+    "                 2 : \"Sports\",\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: You should reference how users can get this information

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "import random\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: I think this is unused

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+    "sub_train_, sub_valid_ = \\\n",
+    "    random_split(train_dataset, [train_len, len(train_dataset) - train_len])\n",
+    "\n",
+    "for epoch in range(N_EPOCHS):\n",
+    "\n",
+    "    start_time = time.time()\n",
+    "    train_loss, train_acc = train_func(sub_train_)\n",
+    "    valid_loss, valid_acc = valid_test_func(sub_valid_)\n",
+    "\n",
+    "    _secs = int(time.time() - start_time)\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: variables with leading underscores are typically reserved for private members of classes. Could we remove the leading underscores here?

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+   "metadata": {},
+   "source": [
+    "## Define functions to train the model and evaluate results."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`torch.utils.data.DataLoader` is recommended for PyTorch users to load data. We use `DataLoader` here to load AG_NEWS datasets and send it to the model for training/validation."
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: Why do we recommend it? Also it could be useful to make this a link into the docs here.

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+    "    train_loss, train_acc = train_func(sub_train_)\n",
+    "    valid_loss, valid_acc = valid_test_func(sub_valid_)\n",
+    "\n",
+    "    _secs = int(time.time() - start_time)\n",
+    "    _mins = _secs / 60\n",
+    "    _secs = _secs % 60\n",
+    "\n",
+    "    print('Epoch: %d' %(epoch + 1), \" | time in %d minutes, %d seconds\" %(_mins, _secs))\n",
+    "    print(f'\\tLoss: {train_loss:.4f}(train)\\t|\\t{valid_loss:.4f}(valid)')\n",
+    "    print(f'\\tAcc: {train_acc * 100:.1f}%(train)\\t|\\t{valid_acc * 100:.1f}%(valid)')\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: This leads to misaligned text. It's a small detail, but could make the output much easier to read.

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+    "        loss.backward()\n",
+    "        optimizer.step()\n",
+    "        train_acc += (output.argmax(1) == cls).sum().item()\n",
+    "\n",
+    "    # Adjust the learning rate\n",
+    "    scheduler.step()\n",
+    "    \n",
+    "    return train_loss / len(sub_train_), train_acc / len(sub_train_)\n",
+    "\n",
+    "def valid_test_func(data_):\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: Could we call this "test" just to be in line with the examples?

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_batch(batch):\n",
+    "    label = torch.tensor([entry[0] for entry in batch])\n",
+    "    text = [entry[1] for entry in batch]\n",
+    "    offsets = [0] + [len(entry) for entry in text]\n",
+    "    offsets = torch.tensor(offsets[:-1]).cumsum(dim=0)\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: I think this call into cumsum can be a bit confusing. It could be good to spell out an example in the text above.

File: docs/tutorials/text_sentiment_ngrams.ipynb [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+    "    Thanks to his best putting performance on the PGA Tour, Rahm \\\n",
+    "    finished with an 8-under 62 for a three-stroke lead, which \\\n",
+    "    was even more impressive considering he’d never played the \\\n",
+    "    front nine at TPC Southwind.\"\n",
+    "\n",
+    "new_label = \"\"\n",
+    "vocab = train_dataset.get_vocab()\n",
+    "with open(\"text_classification.pt\", 'rb') as f:\n",
+    "    model = torch.load(f).to(\"cpu\")\n",
+    "    news_label = ag_news_label[predict(example_text_string,\n",
```
Comment (cpuhrsch) [inline docs/tutorials/text_sentiment_ngrams.ipynb]: If you give "example_text_string" a smaller name (maybe "example") then you can merge this line with the print statement and make this a bit shorter.

**Comment by zhangguanheng66** [comment]
publish on PyTorch Tutorials [here](https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html). Close the PR without merging.
